### DISCUSSION TITLE
QueryClient: stale-while-revalidate support?

### DISCUSSION THREAD

**Discussion by fedebertolini** [TanStack/query#3097] [Q&A] [answered]
Hi team 👋 
Does react-query's `QueryClient` support a stale-while-revalidate query? I see a lot of logic around this is implemented in the basic hook and in the query observer, but maybe I missed something?

Some context and motivation: I'm currently working on a big codebase that was recently moved from a standard express server with SSRing, to NextJS. For an easier migration, the `getServerSideProps` functions were implemented as a set of express-like middlewares that fetch and process data. To be able to cache and share data between different request scopes (only publicly accessible and mostly static data) we implemented this: https://github.com/tannerlinsley/react-query/discussions/998 (global client, with a scoped client proxy that stores the ids of the queries executed by the request to be dehydrated later). In these middlewares we just use the Query Client APIs, mainly `fetchQuery`, `prefetchQuery`, and `setQueryData`.

The problem we have is that when we have **stale** cached data, `fetchQuery` would not return the stale data when the query function throws an error/rejection, it would just return the query's error. Is there a plan to change the behaviour of `fetchQuery` to fallback to stale cached data in case the query fn fails?
Probably this is outside the scope of this project, but it would be nice to have a primitive in the Query Client that would always return the cached data (even if stale) and execute the query function to refresh the data on the background or when there's no data in cache.

I guess I could implement this by using a mix of `getQueryData` to return the cached data, `fetchQuery` when there's no cached data and `refetchQueries` to refresh data on the background.

Any advice would be greatly appreciated 🙏

**Comment by TkDodo** [comment, marked as answer]
I think we're mixing two things here:

1) if you pass `staleTime` to `fetchQuery`, data will be returned if its not older than the passed `staleTime`. This is [documented here](https://react-query.tanstack.com/reference/QueryClient#queryclientfetchquery):

> If the query exists and the data is not invalidated or older than the given staleTime, then the data from the cache will be returned. Otherwise it will try to fetch the latest data.

2) react-query will not remove stale data even in case of error. What you should get is:

```
{
  status: 'error',
  data: theStaleData,
  error: yourError
}
```

if that is not the case, please provide a codesandbox reproduction.

**Reply by fedebertolini** [reply]
Sandbox: https://codesandbox.io/s/nostalgic-bhaskara-js6te

In this sandbox i'm executing `fetchQuery` with the same key twice:
- the first time the query resolves fine
- the second time it throws an error

I'm passing `staleTime` 0 so the data becomes stale right away, and `cacheTime` `Infinity` so the data is always in cache. I would expect `fetchQuery` to return the stale data the second time, but it's throwing the error.

**Reply by TkDodo** [reply]
okay yeah, `fetchQuery` returns a rejected Promise, so you only have the error available in the `catch` clause. I don't see a way how this api can return both error and data - this is different from how useQuery works.

however, data is still in the cache, so you can just do `queryClient.getQueryData(key)` to get the data that's currently stored in the cache, even in the catch clause:

https://codesandbox.io/s/awesome-paper-24vx5?file=/src/index.js

**Reply by fedebertolini** [reply]
Yes, I've been playing around with react query and managed to get something working, but wanted to see if there were any plans to move a bit this logic from the useQuery hook to a query client function, or if there was a better way to implement SWR with query client than using `getQueryData` to return the cached data, `fetchQuery` when there's no cached data and `refetchQueries` to refresh data on the background.

**Reply by fedebertolini** [reply]
I guess another option could be:
1. calling `fetchQuery` passing `staleTime: Infinity` and `cacheTime: Infinity`
2. have a `setInterval` that calls `refetchQueries` for these queries every X minutes

🤔

**Reply by TkDodo** [reply]
I think what you are looking for is what useQuery does without the hook? So I would look towards the `QueryObserver`, which is what we create under the hood in `useQuery`.

**Reply by fedebertolini** [reply]
thanks @TkDodo ! I will take a look at that :)

**Reply by fedebertolini** [reply]
hi @TkDodo, i was just trying out calling `client.refetchQueries` inside a `setInterval` and found an issue with dehydration. If I don't pass a `refetchPage` fn inside the refetch filter options, then the dehydration fails with a serialization error because `query.state.fetchMeta.refetchPage` is `undefined`.
I checked the queryClient code and this is being done in https://github.com/tannerlinsley/react-query/blob/master/src/core/queryClient.ts#L290-L297:
```
      this.queryCache.findAll(filters).map(query =>
        query.fetch(undefined, {
          ...options,
          meta: { refetchPage: filters?.refetchPage },
        })
      )
```
I was not able to reproduce this consistently, maybe it only happens when the dehydration is executed while running the query. I will try passing `refetchPage: null` in the query options, but I guess this react-query shouldn't fail when passing undefined

**Reply by TkDodo** [reply]
I think we should probably not store `refetchPage: undefined` on the `fetchMeta` 🤔

**Reply by fedebertolini** [reply]
my workaround was passing `refetchPage: () => false,` to the `refetchQueries` options, but I agree the query should not enter a state in which dehydration fails.

**Reply by TkDodo** [reply]
> refetchPage: () => false,

`refetchPage: () => true` would be the same as undefined. 

I can take a look into the problem if you can provide a codesandbox reproduction :)
